File: ALTableViewFrameworkExamples/ALTableViewFrameworkExamples/MasterViewController.m
//
//  MasterViewController.m
//  ALTableViewFrameworkExamples
//
//  Created by Abimael Barea Puyana on 20/11/15.
//  Copyright © 2015 Abimael Barea Puyana. All rights reserved.
//

#import "MasterViewController.h"
#import "DetailViewController.h"
#import "MasterTableViewCell.h"
#import "MasterExample1TableViewCell.h"
#import "MasterExample2TableViewCell.h"
#import "MasterExample3TableViewCell.h"
#import "MasterExample4TableViewCell.h"

#import "ALTableViewFrameworkExamples-Swift.h"

@interface MasterViewController ()

@property NSMutableArray *objects;
@end

@implementation MasterViewController


#pragma mark - View LifeCycle

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    self.detailViewController = (DetailViewController *)[[self.splitViewController.viewControllers lastObject] topViewController];
    [self registerCells];
    [self replaceAllSectionElements:[self createElements]];
    [self.tableView setTableFooterView:[[UIView alloc] init]];
}

- (void)viewWillAppear:(BOOL)animated {
    self.clearsSelectionOnViewWillAppear = self.splitViewController.isCollapsed;
    [super viewWillAppear:animated];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


#pragma mark - Register Cells

-(void) registerCells {
    [self registerClass:[MasterTableViewCell class] CellIdentifier:@"MasterTableViewCell"];
}


#pragma mark - Create Cells

- (NSMutableArray *) createElements {
    // Creamos las secciones
    NSMutableArray * sections = [NSMutableArray array];
    
    NSMutableArray * rows = [NSMutableArray array];
    NSNumber * height = [NSNumber numberWithFloat:UITableViewAutomaticDimension];
    
    RowElement * example1 = [RowElement rowElementWithClassName:[MasterExample1TableViewCell class] object:@"Example 1" heightCell:height cellIdentifier:@"MasterTableViewCell"];
    [rows addObject:example1];
    example1.estimateHeightMode = YES;
    
    RowElement * example2 = [RowElement rowElementWithClassName:[MasterExample2TableViewCell class] object:@"Example 2 Index Table View" heightCell:height cellIdentifier:@"MasterTableViewCell"];
    [rows addObject:example2];
    example2.estimateHeightMode = YES;
    
    RowElement * example3 = [RowElement rowElementWithClassName:[MasterExample3TableViewCell class] object:@"Example 3 Twitter Timeline with automatic dimension cells" heightCell:height cellIdentifier:@"MasterTableViewCell"];
    [rows addObject:example3];
    example3.estimateHeightMode = YES;
    
    RowElement * example4 = [RowElement rowElementWithClassName:[MasterExample4TableViewCell class] object:@"Example 4 Swift" heightCell:height cellIdentifier:@"MasterTableViewCell"];
    [rows addObject:example4];
    example4.estimateHeightMode = YES;
    
    [rows addObject:[RowElement rowElementWithClassName:[UITableViewCell class] object:@"Example 5 Swift" heightCell:height cellIdentifier:nil CellStyle:UITableViewCellStyleDefault
                                     CellPressedHandler:^(UIViewController * viewController, UITableViewCell * cell) {
                                         TableViewExample5 * example5 = [[TableViewExample5 alloc] init];
                                         if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad) {
                                             [[((MasterViewController *) viewController) detailViewController] setupViewController:example5];
                                         } else {
                                             [viewController.navigationController pushViewController:example5 animated:YES];
                                         }
                                     } CellCreatedHandler:^(NSString * object, UITableViewCell * cell) {
                                         cell.textLabel.text = object;
                                         cell.textLabel.textAlignment = NSTextAlignmentCenter;
                                     } CellDeselectedHandler:^(UITableViewCell * cell) {
                                         
                                     }]
     ];
    
    SectionElement * sectionElement = [SectionElement sectionElementWithSectionTitleIndex:nil viewHeader:nil viewFooter:nil heightHeader:@0 heightFooter:@0 cellObjects:rows isExpandable:NO];
    
    [sections addObject:sectionElement];
    return sections;
}

@end

File: ALTableViewSwift/ALTableView/MasterViewController.swift
//
//  MasterViewController.swift
//  ALTableView
//
//  Created by lorenzo villarroel perez on 7/3/18.
//  Copyright © 2018 lorenzo villarroel perez. All rights reserved.
//

import UIKit

class MasterViewController: UITableViewController {
    
    let masterTableViewCellString = "MasterTableViewCell"
    let master2TableViewCellString = "Master2TableViewCell"

    var detailViewController: DetailViewController? = nil
    var objects = [Any]()
    var alTableView: ALTableView?

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        if let split = splitViewController {
            let controllers = split.viewControllers
            detailViewController = (controllers[controllers.count-1] as! UINavigationController).topViewController as? DetailViewController
        }
        self.tableView.tableFooterView = UIView()
        
        
        let sectionElements = self.createElements()

        self.alTableView = ALTableView(sectionElements: sectionElements, viewController: self, tableView: self.tableView)
        self.registerCells()
        self.tableView.delegate = self.alTableView
        self.tableView.dataSource = self.alTableView
        self.tableView.reloadData()
    }

    override func viewWillAppear(_ animated: Bool) {
        clearsSelectionOnViewWillAppear = splitViewController!.isCollapsed
        super.viewWillAppear(animated)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Segues

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "showDetail" {
            if let indexPath = tableView.indexPathForSelectedRow {
                let object = objects[indexPath.row] as! NSDate
                let controller = (segue.destination as! UINavigationController).topViewController as! DetailViewController
                controller.detailItem = object
                controller.navigationItem.leftBarButtonItem = splitViewController?.displayModeButtonItem
                controller.navigationItem.leftItemsSupplementBackButton = true
            }
        }
    }
    
    func createElements() -> [ALSectionElement] {
        var sectionElements = [ALSectionElement]()
        for _ in 0...2 {
            var rowElements = Array<ALRowElement>()
            let rowElement = ALRowElement(className:MasterTableViewCell.classForCoder(), identifier: masterTableViewCellString, dataObject: "Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1Texto 1", estimateHeightMode: true)
            let rowElement2 = ALRowElement(className:Master2TableViewCell.classForCoder(), identifier: master2TableViewCellString, dataObject: 12, estimateHeightMode: true)
            rowElements.append(rowElement)
            rowElements.append(rowElement2)
            
            let labelTitle: UILabel = UILabel()
            labelTitle.text = "Header Test"
            labelTitle.backgroundColor = .green
            
            let section = ALSectionElement(rowElements: rowElements, viewHeader: labelTitle, headerHeight: 40, isExpandable: true)
            
            sectionElements.append(section)
        }

        return sectionElements
    }

    func registerCells() {
        self.alTableView?.registerCell(nibName: masterTableViewCellString, reuseIdentifier: masterTableViewCellString)
        self.alTableView?.registerCell(nibName: master2TableViewCellString, reuseIdentifier: master2TableViewCellString)
    }


}

